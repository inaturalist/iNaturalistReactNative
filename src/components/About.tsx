import { useNavigation } from "@react-navigation/native";
import {
  Body1,
  Body2,
  Button,
  Heading4,
  ScrollViewWrapper,
  Tabs,
  UnderlinedLink,
  ViewWrapper
} from "components/SharedComponents";
import { Image, Pressable, View } from "components/styledComponents";
import { t } from "i18next";
import React, { useState } from "react";
import { getBuildNumber, getVersion } from "react-native-device-info";
import { useDebugMode } from "sharedHooks";

const aboutID = "about";
const teamID = "team";

const About = ( ) => {
  const navigation = useNavigation();
  const [activeTab, setActiveTab] = useState( aboutID );
  const [count, setCount] = useState( 0 );
  const appVersion = getVersion();
  const buildVersion = getBuildNumber();
  const { isDebug, toggleDebug } = useDebugMode( );

  const onTermsPressed = async () => {
    const url = "https://www.inaturalist.org/pages/terms";
    navigation.navigate( "FullPageWebView", {
      title: t( "TERMS-OF-USE" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onPrivacyPressed = async () => {
    const url = "https://www.inaturalist.org/pages/privacy";
    navigation.navigate( "FullPageWebView", {
      title: t( "PRIVACY-POLICY" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onCommunityPressed = async () => {
    const url = "https://www.inaturalist.org/pages/community+guidelines";
    navigation.navigate( "FullPageWebView", {
      title: t( "COMMUNITY-GUIDELINES" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onVersionPressed = () => {
    if ( ( count + 1 ) % 3 === 0 ) {
      toggleDebug();
    }
    setCount( count + 1 );
  };

  return (
    <ViewWrapper>
      <Tabs
        tabs={[
          {
            id: aboutID,
            text: t( "ABOUT" ),
            onPress: () => {
              setActiveTab( aboutID );
            }
          },
          {
            id: teamID,
            text: t( "TEAM" ),
            onPress: () => {
              setActiveTab( teamID );
            }
          }
        ]}
        activeId={activeTab}
      />
      <ScrollViewWrapper>
        {activeTab === aboutID && (
          <View className="px-4 py-8">
            <Image
              className="w-[65%] max-h-12 xl:max-h-24 2xl:max-h-32 object-center self-center mb-8"
              resizeMode="contain"
              source={require( "images/inaturalist-dark.png" )}
              accessibilityIgnoresInvertColors
            />
            <Heading4 className="mb-3">
              {t( "INATURALIST-MISSION-VISION" )}
            </Heading4>
            <Body2 className="mb-5">
              {t( "iNaturalist-mission-is-to-connect" )}
            </Body2>
            <Body2 className="mb-8">{t( "iNaturalists-vision-is-a-world" )}</Body2>
            <Heading4 className="mb-3">{t( "WHAT-IS-INATURALIST" )}</Heading4>
            <Body2 className="mb-5">
              {t( "iNaturalist-helps-you-identify" )}
            </Body2>
            <Body2 className="mb-5">{t( "Whats-more-by-recording" )}</Body2>
            <Body2 className="mb-8">{t( "iNaturalist-is-supported-by" )}</Body2>
            <UnderlinedLink
              className="mb-5"
              onPress={() => onTermsPressed()}
            >
              {t( "Terms-of-Use" )}
            </UnderlinedLink>
            <UnderlinedLink
              className="mb-5"
              onPress={() => onPrivacyPressed()}
            >
              {t( "Privacy-Policy" )}
            </UnderlinedLink>
            <UnderlinedLink
              className="mb-8"
              onPress={() => onCommunityPressed()}
            >
              {t( "Community-Guidelines" )}
            </UnderlinedLink>
            <Pressable
              accessibilityRole="button"
              className="items-center justify-center"
              onPress={() => onVersionPressed()}
            >
              <Body1>{ t( "Version-app-build", { appVersion, buildVersion } )}</Body1>
            </Pressable>
            {isDebug && (
              <Button
                text="TURN OFF DEBUG MODE"
                className="mt-5 bg-deeppink"
                level="primary"
                onPress={() => {
                  setCount( 0 );
                  toggleDebug();
                }}
              />
            )}
          </View>
        )}
        {activeTab === teamID && (
          <View className="px-4 py-8">
            <Heading4 className="mb-3">{t( "INATURALIST-TEAM" )}</Heading4>
            <Body2 className="mb-8">
              {t( "iNaturalists-apps-are-designed-and-developed-3" )}
            </Body2>
            <Heading4 className="mb-3">{t( "INATURALIST-COMMUNITY" )}</Heading4>
            <Body2 className="mb-8">
              {t( "iNaturalist-is-supported-by-our-community" )}
            </Body2>
            <Heading4 className="mb-3">{t( "INATURALIST-NETWORK" )}</Heading4>
            <Body2 className="mb-8">{t( "The-iNaturalist-Network" )}</Body2>
            <Heading4 className="mb-3">{t( "COLLABORATORS" )}</Heading4>
            <Body2>{t( "The-models-that-suggest-species" )}</Body2>
          </View>
        )}
      </ScrollViewWrapper>
    </ViewWrapper>
  );
};

export default About;
