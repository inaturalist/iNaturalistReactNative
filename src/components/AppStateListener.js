import { focusManager } from "@tanstack/react-query";
import useDeviceStorageFull from "components/Camera/hooks/useDeviceStorageFull";
import { useEffect, useState } from "react";
import { AppState } from "react-native";
import { log } from "sharedHelpers/logger";
import {
  usePerformance
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

const logger = log.extend( "AppStateListener" );

const AppStateListener = ( ) => {
  const { loadTime } = usePerformance( {
    screenName: "AppStateListener"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }
  const { deviceStorageFull, showStorageFullAlert } = useDeviceStorageFull( );
  const [deviceStorageFullShown, setDeviceStorageFullShown] = useState( false );

  useEffect( ( ) => {
    const subscription = AppState.addEventListener( "change", nextAppState => {
      if ( nextAppState === "active" ) {
        // Check storage and show alert as needed when app becomes active
        if ( deviceStorageFull && !deviceStorageFullShown ) {
          showStorageFullAlert( );
          setDeviceStorageFullShown( true );
        }

        focusManager.setFocused( true );
      }
    } );

    return ( ) => subscription.remove( );
  }, [deviceStorageFull, deviceStorageFullShown, showStorageFullAlert] );

  return null;
};

export default AppStateListener;
