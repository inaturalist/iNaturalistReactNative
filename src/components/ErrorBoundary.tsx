import {
  Body1,
  Button,
  Heading1,
  Heading2
} from "components/SharedComponents";
import {
  SafeAreaView,
  ScrollView
} from "components/styledComponents";
import { t } from "i18next";
import React from "react";
import RNRestart from "react-native-restart";
import { log } from "sharedHelpers/logger";

const logger = log.extend( "ErrorBoundary" );

interface Props {
  children: React.ReactNode;
}

interface State {
  error: Error | null;
  info: React.ErrorInfo | null;
}

// https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary
class ErrorBoundary extends React.Component<Props, State> {
  constructor( props: Props ) {
    super( props );
    this.state = { error: null, info: null };
  }

  static getDerivedStateFromError( error: Error ) {
    // Update state so the next render will show the fallback UI.
    return { error };
  }

  componentDidCatch( error: Error, info: React.ErrorInfo ) {
    this.setState( { info } );
    if ( info?.componentStack ) {
      // componentStack is a little more informative than what's generally in
      // stack, so put that in there before logging
      error.stack = info.componentStack;
    }
    logger.error( error );
  }

  render() {
    if ( this.state.error ) {
      // TODO allow this to take a fallback component as a prop so individual
      // boundaries can customize their error states
      return (
        <SafeAreaView className="flex-1">
          <ScrollView className="p-5 bg-white">
            <Heading1 className="my-3">{ t( "Something-went-wrong" ) }</Heading1>
            <Body1>
              { t( "If-youre-seeing-this-error" ) }
            </Body1>
            <Button
              text={t( "RESTART-APP" )}
              onPress={( ) => RNRestart.restart()}
              className="my-5"
              level="focus"
            />
            <Heading2 className="mt-5">{ t( "Error" ) }</Heading2>
            <Body1>{ this.state.error.toString( ) }</Body1>
            <Body1>{ this.state.info?.componentStack }</Body1>
          </ScrollView>
        </SafeAreaView>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
