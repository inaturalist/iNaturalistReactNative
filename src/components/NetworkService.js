import { log } from "sharedHelpers/logger";
import {
  useIconicTaxa,
  useObservationUpdatesWhenFocused, usePerformance
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

import useTaxonCommonNames from "./hooks/useTaxonCommonNames";
import useWorkQueue from "./hooks/useWorkQueue";

const logger = log.extend( "NetworkService" );

const NetworkService = ( ) => {
  const { loadTime } = usePerformance( {
    screenName: "NetworkService"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }

  useObservationUpdatesWhenFocused( );

  useIconicTaxa( { reload: true } );
  // This only runs when App updates... which is rarely. It works for Settings
  // b/c it generally updates currentUser
  useWorkQueue( );
  useTaxonCommonNames( );

  return null;
};

export default NetworkService;
