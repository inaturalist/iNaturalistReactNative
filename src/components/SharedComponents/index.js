export { default as ActivityAnimation } from "./ActivityAnimation/ActivityAnimation";
export { default as ActivityCount } from "./ActivityCount/ActivityCount";
export { default as CommentsCount } from "./ActivityCount/CommentsCount";
export { default as IdentificationsCount } from "./ActivityCount/IdentificationsCount";
export { default as ActivityIndicator } from "./ActivityIndicator";
export { default as ButtonBar } from "./ButtonBar";
export { default as BackButton } from "./Buttons/BackButton";
export { default as Button } from "./Buttons/Button";
export { default as CloseButton } from "./Buttons/CloseButton";
export { default as EvidenceButton } from "./Buttons/EvidenceButton";
export { default as INatIconButton } from "./Buttons/INatIconButton";
export { default as RotatingINatIconButton } from "./Buttons/RotatingINatIconButton";
export { default as TransparentCircleButton } from "./Buttons/TransparentCircleButton";
export { default as Carousel } from "./Carousel";
export { default as CarouselDots } from "./CarouselDots";
export { default as Checkbox } from "./Checkbox";
export { default as CircleDots } from "./CircleDots";
export { default as ConfidenceInterval } from "./ConfidenceInterval";
export { default as DateDisplay } from "./DateDisplay";
export { default as DateTimePicker } from "./DateTimePicker";
export { default as DisplayTaxon } from "./DisplayTaxon";
export { default as DisplayTaxonName } from "./DisplayTaxonName";
export { default as Divider } from "./Divider/Divider";
export { default as CustomFlashList } from "./FlashList/CustomFlashList";
export { default as CustomRefreshControl } from "./FlashList/CustomRefreshControl";
export { default as FlashListEmptyWrapper } from "./FlashList/FlashListEmptyWrapper";
export { default as InfiniteScrollLoadingWheel } from "./FlashList/InfiniteScrollLoadingWheel";
export { default as FloatingActionBar } from "./FloatingActionBar";
export { default as HideView } from "./HideView";
export { default as IconicTaxonChooser } from "./IconicTaxonChooser";
export { default as IconicTaxonIcon } from "./IconicTaxonIcon";
export { default as INatIcon } from "./INatIcon";
export { default as InlineUser } from "./InlineUser/InlineUser";
export { default as InputField } from "./InputField";
export { default as KebabMenu } from "./KebabMenu";
export { default as KeyboardDismissableView } from "./KeyboardDismissableView";
export { default as LabelColonValue } from "./LabelColonValue";
export { default as DetailsMap } from "./Map/DetailsMap";
export { default as Map } from "./Map/Map";
export { default as MediaNavButtons } from "./MediaNavButtons";
export { default as Modal } from "./Modal";
export { default as Mortal } from "./Mortal";
export { default as ObservationLocation } from "./ObservationLocation";
export { default as ObsStatus } from "./ObsStatus";
export { default as OfflineNotice } from "./OfflineNotice";
export { default as OverlayHeader } from "./OverlayHeader";
export { default as OverviewCounts } from "./OverviewCounts";
export { default as PermissionGate } from "./PermissionGate";
export { default as PhotoCount } from "./PhotoCount";
export { default as QualityGradeStatus } from "./QualityGradeStatus/QualityGradeStatus";
export { default as RadioButtonRow } from "./RadioButtonRow";
export { default as ScrollableWithStickyHeader } from "./ScrollableWithStickyHeader";
export { default as ScrollViewWrapper } from "./ScrollViewWrapper";
export { default as SearchBar } from "./SearchBar";
export { default as BottomSheet } from "./Sheets/BottomSheet";
export { default as BottomSheetStandardBackdrop } from "./Sheets/BottomSheetStandardBackdrop";
export { default as PickerSheet } from "./Sheets/PickerSheet";
export { default as RadioButtonSheet } from "./Sheets/RadioButtonSheet";
export { default as TextInputSheet } from "./Sheets/TextInputSheet";
export { default as TextSheet } from "./Sheets/TextSheet";
export { default as WarningSheet } from "./Sheets/WarningSheet";
export { default as SimpleObservationLocation } from "./SimpleObservationLocation";
export { default as Tabs } from "./Tabs/Tabs";
export { default as TaxonResult } from "./TaxonResult";
export { default as TaxonSearch } from "./TaxonSearch";
export { default as Body1 } from "./Typography/Body1";
export { default as Body2 } from "./Typography/Body2";
export { default as Body3 } from "./Typography/Body3";
export { default as Body4 } from "./Typography/Body4";
export { default as Heading1 } from "./Typography/Heading1";
export { default as Heading2 } from "./Typography/Heading2";
export { default as Heading3 } from "./Typography/Heading3";
export { default as Heading4 } from "./Typography/Heading4";
export { default as Heading5 } from "./Typography/Heading5";
export { default as Heading6 } from "./Typography/Heading6";
export { default as List2 } from "./Typography/List2";
export { default as P } from "./Typography/P";
export { default as Subheading1 } from "./Typography/Subheading1";
export { default as Subheading2 } from "./Typography/Subheading2";
export { default as UnderlinedLink } from "./Typography/UnderlinedLink";
export { default as UploadProgressBar } from "./UploadProgressBar";
export { default as UploadStatus } from "./UploadStatus/UploadStatus";
export { default as UserIcon } from "./UserIcon";
export { default as UserText } from "./UserText";
export { default as ViewWrapper } from "./ViewWrapper";
