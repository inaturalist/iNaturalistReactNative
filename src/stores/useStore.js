import _ from "lodash";
import { MMKV } from "react-native-mmkv";
import { create } from "zustand";
import { createJSONStorage, persist } from "zustand/middleware";

import createExploreSlice from "./createExploreSlice";
import createLayoutSlice from "./createLayoutSlice";
import createMyObsSlice from "./createMyObsSlice";
import createObservationFlowSlice from "./createObservationFlowSlice";
import createRootExploreSlice from "./createRootExploreSlice";
import createSyncObservationsSlice from "./createSyncObservationsSlice";
import createUploadObservationsSlice from "./createUploadObservationsSlice";

export const storage = new MMKV( );

// TODO do *not* export this. This allows any consumer to overwrite *any* part
// of state, circumventing any getter/setter logic we have in the stores. If
// you need to modify state, you should be doing so through a store.
export const zustandStorage = {
  setItem: ( name, value ) => storage.set( name, value ),
  getItem: name => {
    const value = storage.getString( name ) || storage.getNumber( name );
    return value ?? null;
  },
  removeItem: name => storage.delete( name )
};

// Using slices to separate store for Explore and Observation creation flow
// https://docs.pmnd.rs/zustand/guides/slices-pattern
const useStore = create( persist(
  ( ...args ) => {
    // Let's make our slices
    const slices = [
      createExploreSlice( ...args ),
      createLayoutSlice( ...args ),
      createMyObsSlice( ...args ),
      createObservationFlowSlice( ...args ),
      createRootExploreSlice( ...args ),
      createSyncObservationsSlice( ...args ),
      createUploadObservationsSlice( ...args )
    ];

    // Now let's make sure they're not clobbering each other because
    // everything in Zustand state exists in a single namespace and
    // clobbering can happen silently... which is bad.
    const allKeys = slices.map( slice => Object.keys( slice ) ).flat( );
    const keyCounts = allKeys.reduce(
      ( memo, curr ) => {
        memo[curr] ||= 0;
        memo[curr] += 1;
        return memo;
      },
      {}
    );
    const nonUniqueKeys = Object.keys( keyCounts ).reduce(
      ( memo, curr ) => {
        if ( keyCounts[curr] > 1 ) memo.push( curr );
        return memo;
      },
      []
    );
    if ( nonUniqueKeys.length > 0 ) {
      throw new Error(
        `You have multiple Zustand slices with the following keys: ${nonUniqueKeys}`
      );
    }

    // All good? Now let's combine them into one enormous hideous object
    return slices.reduce(
      ( memo, curr ) => ( { ...memo, ...curr } ),
      {}
    );
  },
  {
    name: "persisted-zustand",
    partialize: state => ( {
      // Vestigial un-namespaced values in the layout slice
      isAdvancedUser: state.isAdvancedUser,
      obsDetailsTab: state.obsDetailsTab,

      // Dynamically select all values in the layout slice's namespace
      layout: ( Object.keys( state.layout ).reduce( ( memo, key ) => {
        if ( typeof ( state.layout[key] ) !== "function" ) {
          memo[key] = state.layout[key];
        }
        return memo;
      }, {} ) )
    } ),
    storage: createJSONStorage( () => zustandStorage ),
    // We need to deep merge to persist nested objects, like layout
    // https://zustand.docs.pmnd.rs/middlewares/persist#persisting-a-state-with-nested-objects
    merge: ( persisted, current ) => _.merge( current, persisted )
  }
) );

export default useStore;
