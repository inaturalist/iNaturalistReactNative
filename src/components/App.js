// @flow

import { useNavigation } from "@react-navigation/native";
import RootDrawerNavigator from "navigation/rootDrawerNavigator";
import type { Node } from "react";
import React, { useCallback } from "react";
import { log } from "sharedHelpers/logger";
import {
  useCurrentUser,
  usePerformance,
  useShare
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

import AppStateListener from "./AppStateListener";
import useLinking from "./hooks/useLinking";
import NetworkService from "./NetworkService";
import StartupService from "./StartupService";

const logger = log.extend( "App" );

type SharedItem = {
  mimeType: string,
  data: string | Array<string>
};

const handleShare = ( navigation, item: ?SharedItem ) => {
  if ( !item ) {
    // user hasn't shared any items
    return;
  }

  const { mimeType, data } = item;

  if ( !mimeType && !data ) {
    // user hasn't shared any images
    return;
  }

  // show user a loading animation screen (like PhotoLibrary)
  // while observations are created
  navigation?.navigate( "PhotoSharing", { item } );
};

type Props = {
  // $FlowIgnore
  children?: unknown,
};

// this children prop is here for the sake of testing with jest
// normally we would never do this in code
const App = ( { children }: Props ): Node => {
  const navigation = useNavigation( );
  const { loadTime } = usePerformance( {
    screenName: "App"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }

  // attempting to make sure that navigation is only called once
  // for performance reasons
  const onShare = useCallback(
    item => handleShare( navigation, item ),
    [navigation]
  );

  const currentUser = useCurrentUser( );

  useLinking( currentUser );
  useShare( onShare );

  // this children prop is here for the sake of testing with jest
  // normally we would never do this in code
  return (
    <>
      <StartupService />
      <NetworkService />
      <AppStateListener />
      {children || <RootDrawerNavigator />}
    </>
  );
};

export default App;
