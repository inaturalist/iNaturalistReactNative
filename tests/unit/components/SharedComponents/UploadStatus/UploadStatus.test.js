import { render, screen } from "@testing-library/react-native";
import { UploadStatus } from "components/SharedComponents";
import React from "react";
import useStore from "stores/useStore";

const initialStoreState = useStore.getState( );
beforeAll( ( ) => {
  useStore.setState( initialStoreState, true );
} );

jest.mock( "react-native-circular-progress-indicator", () => "CircularProgress" );

jest.mock( "react-native/Libraries/Animated/NativeAnimatedHelper" );
describe( "UploadStatus", () => {
  it( "displays start icon when upload is unsynced and not queued", async () => {
    render( <UploadStatus progress={0} queued={false} /> );
    expect( screen.getByTestId( /UploadIcon.start/ ) ).toBeTruthy();
    // TODO: (Johannes 2025-06-13) An update to react-native-reanimated (to 3.16.6) caused this test
    // to fail. However, only the snapshot matching is affected, the component still renders
    // correctly, and the test passes, i.e. the correct icon is shown. The actual error thrown is a
    // recursion happening in the pretty-format library. So, since we are currently only half-way
    // through our updating process of major react-native versions and libraries I vouch to comment
    // out these snapshot tests for now and reinstate them later with the latest RN version.
    /*
      RangeError: Invalid string length
      at printObjectProperties (node_modules/pretty-format/build/collections.js:170:47)
      at key (node_modules/pretty-format/build/index.js:386:38)
      at AsymmetricMatcher (node_modules/pretty-format/build/index.js:469:3)
      at printObjectProperties (node_modules/pretty-format/build/collections.js:169:21)
      at key (node_modules/pretty-format/build/index.js:386:38)
      at AsymmetricMatcher (node_modules/pretty-format/build/index.js:469:3)
      at printObjectProperties (node_modules/pretty-format/build/collections.js:169:21)
      at key (node_modules/pretty-format/build/index.js:386:38)
      at AsymmetricMatcher (node_modules/pretty-format/build/index.js:469:3)
    */
    // expect( screen ).toMatchSnapshot();
  } );

  it( "displays rotating circle progress when upload is queued but not started", async () => {
    render( <UploadStatus progress={0} queued /> );
    expect( screen.getByTestId( "UploadStatus.QueuedRotatingIcon" ) ).toBeTruthy();
    // TODO: (Johannes 2025-06-13) An update to react-native-reanimated (to 3.16.6) caused this test
    // to fail. However, only the snapshot matching is affected, the component still renders
    // correctly, and the test passes, i.e. the correct icon is shown. The actual error thrown is a
    // recursion happening in the pretty-format library. So, since we are currently only half-way
    // through our updating process of major react-native versions and libraries I vouch to comment
    // out these snapshot tests for now and reinstate them later with the latest RN version.
    /*
      RangeError: Invalid string length
      at printObjectProperties (node_modules/pretty-format/build/collections.js:170:47)
      at key (node_modules/pretty-format/build/index.js:386:38)
      at AsymmetricMatcher (node_modules/pretty-format/build/index.js:469:3)
      at printObjectProperties (node_modules/pretty-format/build/collections.js:169:21)
      at key (node_modules/pretty-format/build/index.js:386:38)
      at AsymmetricMatcher (node_modules/pretty-format/build/index.js:469:3)
      at printObjectProperties (node_modules/pretty-format/build/collections.js:169:21)
      at key (node_modules/pretty-format/build/index.js:386:38)
      at AsymmetricMatcher (node_modules/pretty-format/build/index.js:469:3)
    */
    // expect( screen ).toMatchSnapshot();
  } );

  it( "displays progress bar when progress is greater than 5%", async () => {
    render( <UploadStatus progress={0.5} /> );
    expect( screen.getByTestId( "UploadStatus.CircularProgress" ) ).toBeTruthy();
    expect( screen ).toMatchSnapshot();
  } );

  it( "displays complete icon when progress is 1", async () => {
    render( <UploadStatus progress={1} /> );
    expect( screen.getByTestId( "UploadStatus.Complete" ) ).toBeTruthy();
    expect( screen ).toMatchSnapshot();
  } );

  it( "has no accessibility errors", () => {
    expect( <UploadStatus progress={0.5} /> ).toBeAccessible();
  } );
} );
