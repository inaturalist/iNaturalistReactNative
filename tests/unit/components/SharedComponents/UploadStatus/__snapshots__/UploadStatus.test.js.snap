// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadStatus displays complete icon when progress is 1 1`] = `
<View
  accessibilityLabel="Upload Complete"
  accessible={true}
  style={
    [
      [
        {
          "justifyContent": "center",
        },
        {
          "height": 65,
        },
      ],
    ]
  }
>
  <View
    style={
      [
        [
          {
            "position": "absolute",
          },
          {
            "height": "100%",
          },
          {
            "justifyContent": "center",
          },
        ],
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "width": 44,
            },
          ],
        ]
      }
    >
      <View
        collapsable={false}
        entering={
          InnerKeyframe {
            "build": [Function],
            "callbackV": undefined,
            "definitions": {
              "0": {
                "opacity": 0,
              },
              "100": {
                "opacity": 0,
              },
              "40": {
                "opacity": 1,
              },
            },
            "delayV": undefined,
            "durationV": 2000,
            "reduceMotionV": "system",
          }
        }
        jestAnimatedStyle={
          {
            "value": {},
          }
        }
        testID="UploadIcon.complete.undefined"
      >
        <View
          style={
            [
              [
                {
                  "alignItems": "center",
                },
                {
                  "justifyContent": "center",
                },
                {
                  "width": 44,
                },
                {
                  "height": 44,
                },
              ],
            ]
          }
          testID="UploadStatus.Complete"
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#74AC00",
                  "fontSize": 28,
                },
                null,
                {
                  "fontFamily": "INatIcon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    entering={
      FadeIn {
        "build": [Function],
        "callbackV": undefined,
        "dampingRatioV": undefined,
        "dampingV": undefined,
        "delayV": 2000,
        "durationV": 1000,
        "easingV": undefined,
        "initialValues": undefined,
        "massV": undefined,
        "overshootClampingV": undefined,
        "randomizeDelay": false,
        "reduceMotionV": "system",
        "restDisplacementThresholdV": undefined,
        "restSpeedThresholdV": undefined,
        "rotateV": undefined,
        "stiffnessV": undefined,
        "type": undefined,
      }
    }
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
  />
</View>
`;

exports[`UploadStatus displays progress bar when progress is greater than 5% 1`] = `
<View
  accessibilityLabel="Upload 50 percent complete"
  accessible={true}
  style={
    [
      [
        {
          "justifyContent": "center",
        },
        {
          "height": 65,
        },
      ],
    ]
  }
>
  <View
    style={
      [
        [
          {
            "width": 44,
          },
        ],
      ]
    }
  >
    <View
      accessibilityLabel="Upload in progress"
      style={
        [
          [
            {
              "alignItems": "center",
            },
            {
              "justifyContent": "center",
            },
            {
              "width": 44,
            },
            {
              "height": 44,
            },
            {
              "justifyContent": "center",
            },
            {
              "alignItems": "center",
            },
          ],
        ]
      }
      testID="UploadIcon.progress.undefined"
    >
      <View
        style={
          [
            [
              {
                "position": "absolute",
              },
            ],
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#454545",
                "fontSize": 15,
              },
              null,
              {
                "fontFamily": "INatIcon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
      <CircularProgress
        activeStrokeColor="#454545"
        activeStrokeWidth={2.5}
        inActiveStrokeOpacity={0}
        maxValue={1}
        radius={18}
        showProgressValue={false}
        testID="UploadStatus.CircularProgress"
        value={0.5}
      />
    </View>
  </View>
</View>
`;

exports[`UploadStatus displays rotating circle progress when upload is queued but not started 1`] = `
<View
  accessibilityLabel="Saved observation, in queue to upload"
  accessible={true}
  style={
    [
      [
        {
          "justifyContent": "center",
        },
        {
          "height": 65,
        },
      ],
    ]
  }
>
  <View
    style={
      [
        [
          {
            "width": 44,
          },
        ],
      ]
    }
  >
    <View
      style={
        [
          [
            {
              "justifyContent": "center",
            },
            {
              "alignItems": "center",
            },
          ],
        ]
      }
    >
      <View
        style={
          [
            [
              {
                "position": "absolute",
              },
            ],
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#454545",
                "fontSize": 15,
              },
              null,
              {
                "fontFamily": "INatIcon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
      <View
        collapsable={false}
        jestAnimatedStyle={
          {
            "value": {
              "transform": [
                {
                  "rotateZ": "0deg",
                },
              ],
            },
          }
        }
        style={
          [
            {
              "transform": [
                {
                  "rotateZ": "0deg",
                },
              ],
            },
          ]
        }
        testID="UploadStatus.QueuedRotatingIcon"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#454545",
                "fontSize": 33,
              },
              null,
              {
                "fontFamily": "INatIcon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`UploadStatus displays start icon when upload is unsynced and not queued 1`] = `
<View
  accessibilityLabel="Saved observation, in queue to upload"
  accessible={true}
  style={
    [
      [
        {
          "justifyContent": "center",
        },
        {
          "height": 65,
        },
      ],
    ]
  }
>
  <View
    style={
      [
        [
          {
            "width": 44,
          },
        ],
      ]
    }
  >
    <View
      accessibilityLabel="Start upload"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "height": 44,
            "width": 44,
          },
          {
            "alignItems": "center",
            "justifyContent": "center",
          },
          false,
          undefined,
          {
            "opacity": 1,
          },
        ]
      }
      testID="UploadIcon.start.undefined"
    >
      <View
        style={
          [
            [
              {
                "position": "relative",
              },
              {
                "left": 0.2,
              },
              {
                "top": 0.1,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "justifyContent": "center",
                },
                {
                  "alignItems": "center",
                },
              ],
            ]
          }
        >
          <View
            style={
              [
                [
                  {
                    "position": "absolute",
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#454545",
                    "fontSize": 15,
                  },
                  null,
                  {
                    "fontFamily": "INatIcon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
          <View
            collapsable={false}
            jestAnimatedStyle={
              {
                "value": {
                  "transform": [
                    {
                      "rotateZ": "0deg",
                    },
                  ],
                },
              }
            }
            style={
              [
                {
                  "transform": [
                    {
                      "rotateZ": "0deg",
                    },
                  ],
                },
              ]
            }
            testID="UploadStatus.QueuedRotatingIcon"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#454545",
                    "fontSize": 33,
                  },
                  null,
                  {
                    "fontFamily": "INatIcon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
